package/src/internal/types.ts

```diff
@@ -0,0 +1,28 @@
+// Source: https://catchts.com/union-array
+export type UnionToIntersection<U> = (
+	U extends any
+		? (k: U) => void
+		: never
+) extends (k: infer I) => void
+	? I
+	: never;
+
+type UnionToOvlds<U> = UnionToIntersection<
+	U extends any ? (f: U) => void : never
+>;
+
+type PopUnion<U> = UnionToOvlds<U> extends (a: infer A) => void ? A : never;
+
+type IsUnion<T> = [T] extends [UnionToIntersection<T>] ? false : true;
```

Comment:
I think that's fine in our case? a Plugin always has at least a name and hook. That being said, I took this type from the article at the top of the file. If you have a better implementation, feel free to suggest it!

Reply:
This is safe for the cases it is currently being used, but I think it might be worth documenting the limitation in case someone else (or future us) try to use this somewhere that breaks.

I don't think there is a guaranteed way to detect whether a given type is a union. Unions can be collapsed and still be equivalent, but would be detected differently here:

```ts
type A = {a: 'a' | 'b'};
type B = {a: 'a'} | {a: 'b'};
```

Those two types are exactly the same, but only B is detected as union. If it gets collapsed to A it is [no longer detected](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgdgSwPZwCpIJJ2BATgZwgGNhk4AeGAPigF4oAKAKAEgYoIAPHOAE3ygBDOCFYsA-IwDWALlgBKOjQBuSBLzFy4EZXiaKuPftLkI4AMzxQMi2irUaJ11lp14A3EyahI1-PDJyVBp6AG1UAF0Obgg+AVCAlHQsHAJiUhQgqijJc0EAG0IoOWBcAFcIT29waABRAEcygvIAQQAaKAAhEKgW6KMBTtZJTv7Y4xaxSVKKzSg8wogmOQXCKurfPvoAb0E5AHJBfagAHyh9gCN9gF9PH2hRnb3zo+vTqF2Dq9uvIhR8YAcRoFOQNJr5Vodbp0KAzSpMP5wAFCRJwOQYfyITItHpwzyI5EXVHozGBaH0PFAA).

Reply:
I think it's safe, I'll add a comment with your example. Thank you!